### PR TITLE
Genode message and shared memory

### PR DESCRIPTION
**Pull request by jklmnn**
- Message and shared memory on Genode
 - Message on Linux
 - Rom client on Linux and Genode
 - Specific asynchronous initialization callbacks
 - Removed Muen support
 - Variant session index type
 - Refactoring of broker

### CODE REVIEW

File: src/message/gneiss-message.ads [hunk 1]
```diff
@@ -12,18 +12,12 @@
...
+   type Message_Buffer is private;
+   Null_Buffer : Message_Buffer;
 package Gneiss.Message with
    SPARK_Mode
 is
-   pragma Compile_Time_Error (Byte'Size /= 8, "Byte size must be 8 bit");
-
-   --  Fixed size message buffer
-   subtype Message_Buffer is Buffer (First .. First + Length - 1);
+   --  pragma Compile_Time_Error (Message_Buffer'Size /= 128 * 8, "Buffer size must be 128 byte");
```
Comment (senier) [inline src/message/gneiss-message.ads]: Why is that commented out? Are we missing a Workarounds entry here?

File: init/gneiss-broker-startup.adb [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+                              Document  :        SXML.Document_Type;
+                              Root      :        SXML.Query.State_Type)
+   is
+      Index   : Positive := Resources'First;
+      State   : SXML.Query.State_Type;
+   begin
+      State := SXML.Query.Path (Root, Document, "/config/resource");
+      while SXML.Query.State_Result (State) = SXML.Result_OK loop
+         State := SXML.Query.Find_Sibling (State, Document, "resource");
+         exit when SXML.Query.State_Result (State) /= SXML.Result_OK;
```
[the remark this answers is not in the document]
Comment (jklmnn) [reply init/gneiss-broker-startup.adb]: The other possible errors are `Result_Invalid` and `Result_Overflow` where from my understanding neither should happen at this point.

File: src/block/gneiss-block.ads [hunk 3]
```diff
@@ -160,55 +157,35 @@ is
    --  Dispatcher capability used to enforce scope for dispatcher session procedures
    type Dispatcher_Capability is limited private;
 
-   --  Return True if C is initialized
-   --
-   --  @param C  Client session instance
-   --  @return   True if initialized
-   function Initialized (C : Client_Session) return Boolean with
+   function Status (Session : Client_Session) return Session_Status with
```
[the remark this answers is not in the document]
Comment (jklmnn) [reply src/block/gneiss-block.ads]: Fixed by https://github.com/Componolit/gneiss/pull/141/commits/69d8000dc1da542ffc34f28c4bbd6c3b6b6b6849.

File: src/block/gneiss-block.ads [hunk 4]
```diff
@@ -160,55 +157,35 @@ is
...
       Annotate => (GNATprove, Terminating);
 
-   --  Return session identifier
-   --
-   --  @param C  Client session instance
-   --  @return   Identifier passed on initialization
-   function Identifier (C : Client_Session) return Session_Id with
-      Annotate => (GNATprove, Terminating),
-      Pre => Initialized (C);
+   function Index (Session : Client_Session) return Session_Index_Option with
```
[the remark this answers is not in the document]
Comment (jklmnn) [reply src/block/gneiss-block.ads]: Fixed by https://github.com/Componolit/gneiss/pull/141/commits/69d8000dc1da542ffc34f28c4bbd6c3b6b6b6849.

File: src/block/gneiss-block.ads [hunk 5]
```diff
@@ -217,13 +194,8 @@ is
...
 
-   --  Return session identifier
-   --
-   --  @param S  Server session instance
-   --  @return   Identifier passed on initialization
-   function Identifier (S : Server_Session) return Session_Id with
-      Annotate => (GNATprove, Terminating),
-      Pre => Valid (S);
+   function Index (Session : Server_Session) return Session_Index_Option with
+      Annotate => (GNATprove, Terminating);
```
[the remark this answers is not in the document]
Comment (jklmnn) [reply src/block/gneiss-block.ads]: Fixed by https://github.com/Componolit/gneiss/pull/141/commits/69d8000dc1da542ffc34f28c4bbd6c3b6b6b6849.

File: src/block/gneiss-block.ads [hunk 6]
```diff
@@ -232,13 +204,8 @@ is
...
 
-   --  Return session identifier
-   --
-   --  @param D  Dispatcher session instance
-   --  @return   Identifier passed on initialization
-   function Identifier (D : Dispatcher_Session) return Session_Id with
-      Annotate => (GNATprove, Terminating),
-      Pre => Initialized (D);
+   function Index (Session : Dispatcher_Session) return Session_Index_Option with
+      Annotate => (GNATprove, Terminating);
```
[the remark this answers is not in the document]
Comment (jklmnn) [reply src/block/gneiss-block.ads]: Fixed by https://github.com/Componolit/gneiss/pull/141/commits/69d8000dc1da542ffc34f28c4bbd6c3b6b6b6849.

File: init/gneiss-broker-main.adb [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+      Parent     : Boolean;
+      Efd        : Gneiss_Epoll.Epoll_Fd;
+   begin
+      Gneiss_Log.Info ("Loading config from " & Conf_Loc);
+      Status           := 1;
+      Conf.Load (Conf_Loc);
+      Query := SXML.Query.Init (State.Xml);
+      if
+         not SXML.Query.Is_Open (Query, State.Xml)
+      then
```
[the remark this answers is not in the document]
Comment (jklmnn) [reply init/gneiss-broker-main.adb]: Fixed by https://github.com/Componolit/gneiss/pull/141/commits/7c1d92d4652ba47b1d08b3c07f8a0d9f3877d49a.

File: src/message/gneiss-message.ads (hunk 1 above)
Comment (jklmnn) [reply src/message/gneiss-message.ads]: I couldn't remember at first what the problem was. Sometimes I have to comment this kind of line out to get more usable error messages in generics. And when testing it locally and on Genode I had no problems anymore enabling this line.
However as you can see in the CI test this error occurs when using GNAT Community 2019 on Linux. When compiling the same code with GNAT Pro 20 it works. It also works on Genode.

Comment (senier) [reply src/message/gneiss-message.ads]: We should create a workaround issue in Componolit/Workarounds with a target tag Community_2020 (which hopefully fixes the issue) and comment out the check again.

**Comment by senier** [comment]
Remove `test/message_server/print.c`

File: src/message/gneiss-message.ads (hunk 1 above)
Comment (jklmnn) [reply src/message/gneiss-message.ads]: Documented in https://github.com/Componolit/Workarounds/issues/18 and fixed in https://github.com/Componolit/gneiss/pull/141/commits/b872839a0bb921cbbe07c90949941f9c8083172d.

**Comment by jklmnn** [comment]
> Remove `test/message_server/print.c`

Fixed in https://github.com/Componolit/gneiss/pull/141/commits/109a4512674502b9f6f3e4fc02ebf2eae35f4fe1.

**Pull request merged**